lib/render/ParticleRenderer.cpp

```diff
@@ -52,96 +55,420 @@ using namespace VAPoR;
 static RendererRegistrar<ParticleRenderer> registrar(ParticleRenderer::GetClassType(), ParticleParams::GetClassType());
 
 ParticleRenderer::ParticleRenderer(const ParamsMgr *pm, string winName, string dataSetName, string instName, DataMgr *dataMgr)
-: Renderer(pm, winName, dataSetName, ParticleParams::GetClassType(), ParticleRenderer::GetClassType(), instName, dataMgr)
+: Renderer(pm, winName, dataSetName, ParticleParams::GetClassType(), ParticleRenderer::GetClassType(), instName, dataMgr),
+  _colorMapTexOffset(0)
 {
+    _cacheParams.ts=0;
+    _cacheParams.rLevel=0;
+    _cacheParams.cLevel=0;
+    _cacheParams.stride=0;
+    _cacheParams.radius=8.0;
+    _cacheParams.directionScale=0.;
+    _cacheParams.direction=false;
 }
 
-ParticleRenderer::~ParticleRenderer() {}
+ParticleRenderer::~ParticleRenderer() {
+    if (_vertexArrayId) {
+        glDeleteVertexArrays(1, &_vertexArrayId);
+        _vertexArrayId = 0;
+    }
+    if (_vertexBufferId) {
+        glDeleteBuffers(1, &_vertexBufferId);
+        _vertexBufferId = 0;
+    }
+
+    if (_colorMapTexId) {
+        glDeleteTextures(1, &_colorMapTexId);
+        _colorMapTexId = 0;
+    }
+}
 
 int ParticleRenderer::_paintGL(bool)
 {
-    int rc = 0;
+#ifdef DEBUG
+    auto start = chrono::steady_clock::now();
+#endif
+
     glDepthMask(true);
     glEnable(GL_DEPTH_TEST);
 
+    if (_particleCacheIsDirty()) {
+        _resetParticleCache();
+    }
+
+    if (_colormapCacheIsDirty()) {
+        _resetColormapCache();
+        _prepareColormap();
+    }
+
+    Grid* grid = nullptr;
+    std::vector<Grid*> vecGrids;
+    int rc = _getGrids(grid, vecGrids);
+    if (rc != 0) {
+        SetErrMsg("Could not get scalar and field grids for ParticleRenderer");
+        return rc;
+    }
+
+    bool renderGlyphs = GetActiveParams()->GetValueLong(ParticleParams::Render3DTag, true);
+    if (renderGlyphs) {
+        _generateParticleGlyphs(grid, vecGrids);
+        _renderParticleGlyphs();
```

Comment:
It appears the particles are no longer being cached.

Reply:
You're right - thanks for that and fixed.

Reply:
They're still not being cached.